plugins/cr8-copy-s3/src/main/java/io/crate/copy/s3/S3FileOutput.java

```diff
@@ -50,8 +52,15 @@
 public class S3FileOutput implements FileOutput {
 
     @Override
-    public OutputStream acquireOutputStream(Executor executor, URI uri, WriterProjection.CompressionType compressionType) throws IOException {
-        OutputStream outputStream = new S3OutputStream(executor, uri, new S3ClientHelper());
+    public OutputStream acquireOutputStream(Executor executor,
+                                            URI uri,
+                                            WriterProjection.CompressionType compressionType,
+                                            @Nullable String protocolSetting) throws IOException {
+        String fileUri = uri.toString();
+        if (fileUri.startsWith("s3") || fileUri.startsWith("S3")) {
+            fileUri = URIHelper.convertToURI(fileUri);
+        }
```

Comment:
Could we have a single `uri -> S3Uri` conversion instead of going from `String` to `URI` and then later on requiring a `S3ClientHelper.PathParser` ?

Would make it easier to follow which "state" the URI is in. 
See also https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

Reply:
Hi @mfussenegger if I understood correctly, the same should happen on `S3Input` side and the flow from `FileReadingIterator` to `FileInput.listUris` and `FileInput.getStream` will be affected. Since URI is a final class I think this change will be something larger than what it could achieve.
Another option could be to wrap the S3 client methods that we currently use (ex. `listObjects`, `getObject`) to accept URI as an argument (`pathParser` called internally). That way, as a caller you do not need to know the state of the URI.

Reply:
Thank you for suggestion, could you have a quick look at 0703dd3 and confirm if I understood you correctly?

Reply:
No this goes a bit in the wrong direction. The `server` module shouldn't contain any `s3` specific knowledge.


I think that we parse parses the string uri into a `URI` within the `server` module is fine and necessary to get the scheme and select the right factory from the `CopyPlugin`.

Any post-processing should then happen in the scheme specific components and ideally get reflected with type changes.


At which point do we need to handle the host/bucket name? What I'd like to avoid is that we re-create another `URI` that is structured differently.

This PR had some kind of `String -> URI -> URI` or `String -> URI -> String` flow. What I had in mind is to change it to `String -> URI -> SomethingElse` where the last part happens internally in the s3 specific module.

If `listUris` and `getStream` both need this processed URI we could change the `FileInput` interface:


```
FileInput<T>

  T parseURI(URI)
  List<T> listURIs(...)
  InputStream getStream(T)
```

Or we just re-parse the URI in both `listURIs` and `getStream` and keep the `URI` within the interface.

Reply:
@mfussenegger, could you have another look at bab1b86?
I aimed for your second option:
> Or we just re-parse the URI in both listURIs and getStream and keep the URI within the interface.

If this is not what you meant, I would hope to have a quick chat to clarify. I think my initial commit was basically the second option with the parser being part of the server module.

Reply:
Changes like in https://github.com/crate/crate/blob/bab1b86dafd76a70e6c7783569a7bdb5ffa960ac/server/src/main/java/io/crate/execution/engine/export/FileWriterCountCollector.java#L101-L103

would still be a problem. Nothing in `server` should know anything specific about `s3` 
The `CopyPlugin` interface should expose everything that is needed for a plugin to cover all `s3` specific implementation details.

Reply:
Technically, I should be able to drop the condition, `if (uriStr.startsWith("s3") || uriStr.startsWith("S3")) { `
I just had it there to avoid touch non-s3 URIs. Will proceed to remove it. Thank you

Reply:
Please have a look at b26a7bb. This also exported `S3ClientHelper`. Thank you

Reply:
@mfussenegger Please have a quick look at 966728d. I exported the mentioned logic and `s3ClientHelper` as well.